File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/thundercoreV2/utils/Logger.java
package org.firstinspires.ftc.teamcode.thundercoreV2.utils;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;

import org.firstinspires.ftc.teamcode.blacksmithcore.Path;
import org.firstinspires.ftc.teamcode.blacksmithcore.Vector2d;

public class Logger {

    public static Logger instance = new Logger();
    public static Logger getInstance() { return instance; }

    private TelemetryPacket currentPacket = new TelemetryPacket();
    private final FtcDashboard dashboard = FtcDashboard.getInstance();

    private Logger() { }

    public void update() {
        dashboard.sendTelemetryPacket(currentPacket);
        currentPacket = new TelemetryPacket();
    }

    public void put(String name, Object val) {
        currentPacket.put(name, val);
    }

    public void drawVector(Vector2d vector, Vector2d position, String color) {
        double arrowAngle = Math.PI + vector.angle();
        double arrowLen = vector.magnitude()/10;
        Vector2d p2 = position.plus(vector);
        currentPacket.fieldOverlay()
                .setStroke(color)
                .setStrokeWidth(1)
                .strokeLine(position.getX(), position.getY(), p2.getX(), p2.getY())
                .strokeLine(p2.getX(), p2.getY(), p2.getX() + arrowLen*Math.cos(arrowAngle+0.35), p2.getY() + arrowLen*Math.sin(arrowAngle+0.35))
                .strokeLine(p2.getX(), p2.getY(), p2.getX() + arrowLen*Math.cos(arrowAngle-0.35), p2.getY() + arrowLen*Math.sin(arrowAngle-0.35));

    }

    public void drawRobot(Vector2d position, double heading, String color, Double radius) {
        currentPacket.fieldOverlay()
                .setStroke(color)
                .setStrokeWidth(1)
                .strokeLine(position.getX(), position.getY(), position.getX() + radius*Math.cos(heading), position.getY() + radius*Math.sin(heading))
                .strokeCircle(position.getX(), position.getY(), radius);
    }

    public void drawPath(Path path, String color) {
        int res = Math.max((int)path.getLength() / 5, 10);
        double[] xPoints = new double[res+1];
        double[] yPoints = new double[res+1];
        for (int i = 0; i <= res; i++) {
            Vector2d point = path.atParameter(i/(double)res).getVector();
            xPoints[i] = point.getX();
            yPoints[i] = point.getY();
        }
        currentPacket.fieldOverlay()
                .setStroke(color)
                .setStrokeWidth(1)
                .strokePolyline(xPoints, yPoints);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/thundercore/utils/Logger.kt
package org.firstinspires.ftc.teamcode.thundercore.utils

import android.util.Log
import com.acmerobotics.dashboard.FtcDashboard
import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import org.firstinspires.ftc.robotcore.external.Telemetry
import org.firstinspires.ftc.teamcode.blacksmithcore.*
import kotlin.math.*

object Logger {
    private val dashboard = FtcDashboard.getInstance()
    var telemetry: Telemetry? = null
    var configuration = LoggerConfiguration.DASHBOARD_AND_TELEMETRY

    private val valuesToLog = HashMap<String, Any?>()
    private var packet: TelemetryPacket = TelemetryPacket()


    fun put(name: String, value: Any?) {
        if (!configuration.dashboardEnabled && !configuration.telemetryEnabled) return // No point in tracking this
        valuesToLog[name] = value
    }

    fun logMessage(message: String) {
        if (!configuration.androidLoggerEnabled) return // Log disabled
        Log.d("ThunderCore", message)
    }

    fun logWarning(message: String) {
        if (!configuration.androidLoggerEnabled) return // Log disabled
        Log.w("ThunderCore", message)
    }

    fun drawRobot(pose: Pose2d, trackWidth: Double, wheelBase: Double) {
        packet.fieldOverlay()
            .setStroke("green")
            .strokePolygon(doubleArrayOf(
                    pose.x - wheelBase*cos(pose.theta) - trackWidth*sin(pose.theta),
                    pose.x + wheelBase*cos(pose.theta) - trackWidth*sin(pose.theta),
                    pose.x + wheelBase*cos(pose.theta) + trackWidth*sin(pose.theta),
                    pose.x - wheelBase*cos(pose.theta) + trackWidth*sin(pose.theta)
            ), doubleArrayOf(
                    pose.y - wheelBase*sin(pose.theta) + trackWidth*cos(pose.theta),
                    pose.y + wheelBase*sin(pose.theta) + trackWidth*cos(pose.theta),
                    pose.y + wheelBase*sin(pose.theta) - trackWidth*cos(pose.theta),
                    pose.y - wheelBase*sin(pose.theta) - trackWidth*cos(pose.theta),
            ))
            .strokeLine(pose.x, pose.y, pose.x+wheelBase*cos(pose.theta), pose.y+wheelBase*sin(pose.theta))
    }

    fun drawRobot(pose: MutablePose2D, trackWidth: Double, wheelBase: Double) {
        drawRobot(pose.toPose(), trackWidth, wheelBase)
    }

    fun drawTrajectory(trajectory: Trajectory, res: Int = 30) {
        val (xData, yData) = linspace(0.0, trajectory.getLength(), res).map {
            trajectory(it).let { point -> Pair(point.x.x, point.y.x) }
        }.unzip()
        packet.fieldOverlay()
            .setStroke("blue")
            .strokePolyline(xData.toDoubleArray(), yData.toDoubleArray())
    }

    /** Draws a vector on the field */
    private fun drawVector(vector: Vector2d, start: Vector2d, color: String) {
        val theta = atan2(vector.y, vector.x)
        val rho = hypot(vector.y, vector.x)

        packet.fieldOverlay()
            .setStroke(color)
            .strokeLine(start.x, start.y, start.x + vector.x, start.y + vector.y)
            .strokePolyline(
                doubleArrayOf(
                    start.x + 0.9*rho*cos(theta + PI/12),
                    start.x + vector.x,
                    start.x + 0.9*rho*cos(theta - PI/12)
                ), doubleArrayOf(
                    start.y + 0.9*rho*sin(theta + PI/12),
                    start.y + vector.y,
                    start.y + 0.9*rho*sin(theta - PI/12)
                )
            )
    }

    /** Draws an red velocity vector and an orange acceleration vector on the robot. */
    fun drawForceVectors(vector: Vector2dDual) {
        drawVector(Vector2d(vector.x.x, vector.y.x), Vector2d(vector.x.dx, vector.y.dx), "red");
        drawVector(Vector2d(vector.x.x, vector.y.x), Vector2d(vector.x.ddx, vector.y.ddx), "orange");
    }

    fun update() {
        if (!configuration.telemetryEnabled && !configuration.dashboardEnabled) return
        for ((key, value) in valuesToLog) {
            if (configuration.telemetryEnabled) telemetry?.addData(key, value)
            if (configuration.dashboardEnabled) packet.put(key, value)
        }

        if (configuration.telemetryEnabled) telemetry?.update()
        if (configuration.dashboardEnabled) dashboard.sendTelemetryPacket(packet)
        packet = TelemetryPacket()
        valuesToLog.clear()
    }
}